services/src/main/java/org/keycloak/authentication/authenticators/browser/ConditionalOtpFormAuthenticator.java

```diff
@@ -266,4 +269,22 @@ private boolean userHasRole(AuthenticationFlowContext context, String roleName)
 
         return hasRole(user.getRoleMappings(), role);
     }
+
+    private boolean isOTPRequired(RealmModel realm, UserModel user) {
+        for (AuthenticatorConfigModel configModel : realm.getAuthenticatorConfigs()) {
+            String config = configModel.getConfig().get(FORCE_OTP_ROLE);
+            if (config != null) {
+                RoleModel role = getRoleFromString(realm, config);
+                return hasRole(user.getRoleMappings(), role);
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
+        if(isOTPRequired(realm, user) && !user.getRequiredActions().contains(UserModel.RequiredAction.CONFIGURE_TOTP.name())) {
+            user.addRequiredAction(UserModel.RequiredAction.CONFIGURE_TOTP.name());
```

Comment:
What if the config is that a user can access without OTP on an internal machine (based on header), but not remotely. Then he by accident tries to login from a remote machine and is displayed OTP form? As the required action is added to the user here he'll have to setup OTP on the internal machine as well. It should be possible to add the action to the client session instead which would prevent it from being required on a new session.